trio/_core/_run.py

```diff
@@ -1092,7 +1100,71 @@ class _RunStatistics:
     run_sync_soon_queue_size = attr.ib()
 
 
-@attr.s(eq=False, hash=False)
+# This holds all the state that gets trampolined back and forth between
+# callbacks when we're running in guest mode.
+#
+# It has to be a separate object from Runner, and Runner *cannot* have hold
+# references to it (directly or indirectly)!
+#
+# The idea is that we want a chance to detect if our host loop quits and stops
+# driving us forward. We detect that by unrolled_run_gen being garbage
+# collected, and hitting its 'except GeneratorExit:' block. So this only
+# happens if unrolled_run_gen is GCed.
+#
+# The Runner state is referenced from the global GLOBAL_RUN_CONTEXT. The only
+# way it gets *un*referenced is by unrolled_run_gen completing, e.g. by being
+# GCed. But if Runner has a direct or indirect reference to it, and the host
+# loop has abandoned it, then this will never happen!
+#
+# So this object can reference Runner, but Runner can't reference it. The only
+# references to it are the "in flight" callback chain on the host loop /
+# worker thread.
+@attr.s(eq=False, hash=False, slots=True)
+class GuestState:
+    runner = attr.ib()
+    run_sync_soon_threadsafe = attr.ib()
+    done_callback = attr.ib()
+    unrolled_run_gen = attr.ib()
+    unrolled_run_next_send = attr.ib(factory=lambda: Value(None))
```

Comment:
`Outcome.send` and `Outcome.unwrap` are mutators – they consume the object. (To avoid accidentally rethrowing the same exception twice and getting garbled tracebacks.)

Reply:
Right you are, my bad!